File: src/Azdo.Core/Util/Browser.cs
using System.Diagnostics;
using System.Runtime.InteropServices;

namespace Azdo.Core.Util;

/// <summary>Opens URLs in the user's default browser across platforms (≈ <c>browser.Open</c>).</summary>
public static class Browser
{
    public static void Open(string rawUrl)
    {
        if (string.IsNullOrEmpty(rawUrl))
            throw new ArgumentException("browser: empty URL");

        if (!Uri.TryCreate(rawUrl, UriKind.Absolute, out var uri))
            throw new ArgumentException($"browser: invalid URL: {rawUrl}");
        if (uri.Scheme != "http" && uri.Scheme != "https")
            throw new ArgumentException($"browser: only http/https URLs are supported, got \"{uri.Scheme}\"");
        if (string.IsNullOrEmpty(uri.Host))
            throw new ArgumentException($"browser: URL missing host: \"{rawUrl}\"");

        var (name, args) = PlatformCommand(rawUrl);
        Process.Start(new ProcessStartInfo
        {
            FileName = name,
            Arguments = string.Join(' ', args.Select(a => a.Contains(' ') ? $"\"{a}\"" : a)),
            UseShellExecute = false,
        });
    }

    private static (string, string[]) PlatformCommand(string url)
    {
        if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX)) return ("open", new[] { url });
        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows)) return ("rundll32", new[] { "url.dll,FileProtocolHandler", url });
        return ("xdg-open", new[] { url });
    }
}

File: reference/internal/browser/browser.go
// Package browser opens URLs in the user's default web browser across
// macOS, Linux, and Windows. It hands off to the OS handler so whichever
// browser the user has configured as default (Firefox, Chrome, Safari,
// Edge, etc.) is what opens the URL.
package browser

import (
	"fmt"
	"net/url"
	"os/exec"
	"runtime"
)

// execCommand is a package-level seam so tests can intercept the launcher
// invocation without actually spawning a browser.
var execCommand = exec.Command

// Open launches the given URL in the user's default browser. The call
// returns as soon as the launcher process has been started — it does not
// wait for the browser to finish opening.
func Open(rawURL string) error {
	if rawURL == "" {
		return fmt.Errorf("browser: empty URL")
	}

	parsed, err := url.Parse(rawURL)
	if err != nil {
		return fmt.Errorf("browser: invalid URL: %w", err)
	}
	if parsed.Scheme != "http" && parsed.Scheme != "https" {
		return fmt.Errorf("browser: only http/https URLs are supported, got %q", parsed.Scheme)
	}
	if parsed.Host == "" {
		return fmt.Errorf("browser: URL missing host: %q", rawURL)
	}

	name, args := platformCommand(rawURL)
	cmd := execCommand(name, args...)
	if err := cmd.Start(); err != nil {
		return fmt.Errorf("browser: failed to launch %s: %w", name, err)
	}
	return nil
}

// platformCommand returns the OS-appropriate launcher command for opening
// a URL with the user's default browser.
func platformCommand(rawURL string) (string, []string) {
	switch runtime.GOOS {
	case "darwin":
		return "open", []string{rawURL}
	case "windows":
		return "rundll32", []string{"url.dll,FileProtocolHandler", rawURL}
	default:
		return "xdg-open", []string{rawURL}
	}
}
